### PR TITLE
feat: `generateState()` and `generateChallenge()`

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

- implement `generateState()`
- implement `generateCodeChallenge()` (resolves LOG-953)
- generate URL safe strings for Code Verifier and State

## Testing

UT

## Reviewers

@logto-io/eng

### CODE REVIEW

File: Sources/Logto/Extensions/String.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import Foundation
+
+extension String {
+    static func fromUrlSafeBase64(string: String) -> String? {
+        guard let data = Data(base64Encoded: string
+            .replacingOccurrences(of: "-", with: "+")
+            .replacingOccurrences(of: "_", with: "/")
+            // Base64 encoding
+            .padding(toLength: string.count + string.count % 3, withPad: "=", startingAt: 0))
```
Comment (simeng-li) [inline Sources/Logto/Extensions/String.swift:16]: just wondering do we need to align the padding "="  chars in js SDK as well, current `UrlSafeBase64` methods defined in silverhand essentials does not append those padding "=" chars

**Review by simeng-li — APPROVED** [review]
overall LGTM

**Review by wangsijie — APPROVED** [review]
LGTM

File: Sources/Logto/Extensions/String.swift (hunk 1 above)
Comment (gao-sun) [reply Sources/Logto/Extensions/String.swift]: actually this is a reverse action. we remove "="s on purpose when encoding to url-safe strings.

**Review by wangsijie — APPROVED** [review]
LGTM

**Review by simeng-li — APPROVED** [review]

**Pull request merged**